subsys/mgmt/hawkbit/hawkbit.c

```diff
@@ -0,0 +1,1267 @@
+		close.id = acid;
+		close.time = "";
+		close.status.execution = exec;
+		close.status.result.finished = fini;
+
+		ret = json_obj_encode_buf(json_close_descr,
+					  ARRAY_SIZE(json_close_descr), &close,
+					  hb_context.status_buffer,
+					  hb_context.status_buffer_size - 1);
+		if (ret) {
+			LOG_ERR("Can't encode the json script");
+			return false;
+		}
+
+		hb_context.http_req.content_type_value =
+			HTTP_HEADER_CONTENT_TYPE_JSON;
+		hb_context.http_req.payload = hb_context.status_buffer;
+		hb_context.http_req.payload_len =
+			strlen(hb_context.status_buffer);
+
+		ret = http_client_req(hb_context.sock, &hb_context.http_req,
+				      HAWKBIT_RECV_TIMEOUT, "HAWKBIT_CLOSE");
+		if (ret < 0) {
+			LOG_ERR("Unable to send http request");
+			return false;
+		}
+
+		break;
+
+	case HAWKBIT_PROBE_DEPLOYMENT_BASE:
+		hb_context.http_req.content_type_value = NULL;
+		ret = http_client_req(hb_context.sock, &hb_context.http_req,
+				      HAWKBIT_RECV_TIMEOUT,
+				      "HAWKBIT_PROBE_DEPLOYMENT_BASE");
+		if (ret < 0) {
+			LOG_ERR("Unable to send http request");
+			return false;
+		}
+
+		break;
+
+	case HAWKBIT_REPORT:
+		if (!fini || !exec) {
+			return -EINVAL;
+		}
+
+		LOG_INF("Reporting deployment feedback %s (%s) for action %d",
+			fini, exec, hb_context.json_action_id);
+		/* Build JSON */
+		memset(&feedback, 0, sizeof(feedback));
+		snprintk(acid, sizeof(acid), "%d", hb_context.json_action_id);
+		feedback.id = acid;
+		feedback.status.result.finished = fini;
+		feedback.status.execution = exec;
+
+		ret = json_obj_encode_buf(json_dep_fbk_descr,
+					  ARRAY_SIZE(json_dep_fbk_descr),
+					  &feedback, hb_context.status_buffer,
+					  hb_context.status_buffer_size - 1);
+		if (ret) {
+			LOG_ERR("Can't encode response: %d", ret);
+			return ret;
+		}
+
+		hb_context.http_req.content_type_value =
+			HTTP_HEADER_CONTENT_TYPE_JSON;
+		hb_context.http_req.payload = hb_context.status_buffer;
+		hb_context.http_req.payload_len =
+			strlen(hb_context.status_buffer);
+
+		ret = http_client_req(hb_context.sock, &hb_context.http_req,
+				      HAWKBIT_RECV_TIMEOUT, "HAWKBIT_REPORT");
+		if (ret < 0) {
+			LOG_ERR("Unable to send http request");
+			return false;
+		}
+
+		break;
+
+	case HAWKBIT_DOWNLOAD:
+		ret = http_client_req(hb_context.sock, &hb_context.http_req,
+				      HAWKBIT_RECV_TIMEOUT, "HAWKBIT_DOWNLOAD");
+		if (ret < 0) {
+			LOG_ERR("Unable to send image download request");
+			return false;
+		}
+
+		break;
+	}
+
+	return true;
+}
+
+enum hawkbit_response hawkbit_probe(void)
+{
+	int ret;
+	int32_t action_id;
+	int32_t file_size = 0;
+	char device_id[DEVICE_ID_HEX_MAX_SIZE] = { 0 },
+	     cancel_base[CANCEL_BASE_SIZE] = { 0 },
+	     download_http[DOWNLOAD_HTTP_SIZE] = { 0 },
+	     deployment_base[DEPLOYMENT_BASE_SIZE] = { 0 },
+	     firmware_version[BOOT_IMG_VER_STRLEN_MAX] = { 0 };
+
+	memset(&hb_context, 0, sizeof(hb_context));
+	hb_context.response_data = malloc(RESPONSE_BUFFER_SIZE);
```

Comment:
You should verify that `response_data` is not NULL, in case the allocation failed.